### DISCUSSION TITLE
[Walkthrough] Earth 3: Making and Coordinating Bots

### DISCUSSION THREAD

**Discussion by cbot-coder** [colobot/colobot#1376] [Workshop]
## Overview

This mission is still fairly simple, but it does introduce several new things:
- Use of a `class` to encapsulate the solution
- Researching new technologies
- Making a new bot (including passing it code to run)
- Powering up the new bot
- Coordination of roles between the two bots
- Fetching of resources, including the Black Box
- Random quirks in the built-in `goto()` that must be overcome

## Code and Explanations

Let's dive in. The code that gets it all started is short:

```c++
extern void object::SolveEarth3() {

  message("Please get on the ship. Bots will do the work for you.");

  Earth3Solution solution(this);
}
```

What is this doing? It creates an object called `solution`, an instance of the class `Earth3Solution`, and passes `this` (a reference to the bot itself). The `Earth3Solution` class takes care of everything from there.

Here is the beginning of the class, including everything up to the constructor:

```c++
// Keep the solution inside a single class
public class Earth3Solution {
  // References to various objects
  private object bot, ship, rcenter, bfactory, blackbox;

  // Constructor
  void Earth3Solution(object bot) {
    this.bot = bot;

    // Locate various objects
    ship = search(SpaceShip);
    rcenter = search(ResearchCenter);
    bfactory = search(BotFactory);
    blackbox = search(BlackBox);

    // Run code according to the bot type
    if (bot.category == WheeledGrabber) {
      step1();
      step2();
      waitForFinish();
    } else if (bot.category == TrackedGrabber) {
      step3();
    }
  }
```

Notice we keep references to various objects used in this mission; this is just to make code simpler. In the constructor, we first save a reference to the bot object that is running the code; this is critical. We also find all the other objects of interest.

In the constructor, which is called by both bots, we distinguish which parts of the solution will be run by each bot based on its type. This was one of my earlier solutions, so I broke it down into numbered steps. Starting with Tropica I switched to a different approach, but for now this works. The actual steps are then defined in methods which are still members of the `Earth3Solution` class:

```c++
  // Step 1: Research tracks
  public void step1() {
    // ...only if not already done, of course
    if (not researchDone(ResearchTracked)) {
      // Power the Research Center and do the research
      if (rcenter.energyCell == null or rcenter.energyCell.energyLevel < 1) {
        fetch(PowerCell);
        rgoto(rcenter);
        if (rcenter.energyCell != null) {
          drop(Behind); grab();
          turn(-90); drop();
          turn(90); grab(Behind);
        }
        drop();
      }
      rcenter.research(ResearchTracked);
    }
  }
```

Note some of the design principles I laid out in my first walkthrough post at https://github.com/colobot/colobot/discussions/1375: it checks if the research is done first, so as not to repeat it if the program happens to be run twice. It calls a helper function `researchDone()` (to be defined later) to check. If not done, it calls another helper function `fetch()` which takes the category of resource to fetch. Then it goes to the Research Center, drops the cell (after first removing any old cell if present), then asks for the research to start.

```c++
  // Step 2: Create the tracked grabber bot
  public void step2() {
    // ...only if not already done, of course
    object bot2 = search(TrackedGrabber);
    if (bot2 == null) {
      string bot2Program = "" +
        "extern void object::BlackBox2Ship() {\n" +
        "  Earth3Solution solution(this);\n" +
        "}\n";

      // Put Titanium in the Bot Factory
      fetch(Titanium);
      rgoto(bfactory);
      drop(); move(-3);

      // Wait for research to be done, then make the bot
      while (not researchDone(ResearchTracked)) wait(1);
      bfactory.factory(TrackedGrabber, bot2Program);
    }

    // Does the new bot need a battery?
    if (bot2 == null or bot2.energyCell == null) {
      fetch(PowerCell);
      if (bot2 == null) {
        // It's still being made; wait for it
        rgoto(bfactory);
        while((bot2 = radar(TrackedGrabber, 0, 45, 0, 10)) == null or
              bfactory.busy()) wait(1);
        wait(3);
      }
      // Power it up
      rgoto(bot2); drop(); move(-5);
    }

    message("Tracked grabber is ready to go on its mission.");
  }
```

There's a lot going on here. Again, the code checks if there is already a `TrackedGrabber` and skips making one if there is. If it needs to make one, it finds the titanium, drops it in the Bot Factory, waits in case the `Tracked` research is not yet done, then asks the factory to make a new `TrackedGrabber` bot.

The second thing it passes to the bot factory is the program code for the new bot. If you look carefully, it is a string representation of an `extern` wrapper function that does the same thing as the first section of code above, the `SolveEarth3` function: it creates a new `Earth3Solution` object and passes the `this` reference (of the new bot) into it. When it runs, it goes through the same process as the first bot, except that the constructor will identify it as a `TrackedGrabber` and run the `TrackedGrabber`'s code instead!

Finally, it finds a PowerCell and drops it on the new bot (after waiting for it to finish being made). Note the conditions which are carefully crafted to check and wait until the new bot is ready. The extra `wait(3)` is to give some margin time for things to finish. In my experience, sometimes it works without it, but there were times that rushing caused problems.

```c++
  // This is the last step for the wheeled grabber
  public void waitForFinish() {
    object guy = search(Me), bot2 = search(TrackedGrabber);
    rgoto(ship);

    // Wait for everyone to get on the ship
    while (distance2d(bot2.position, ship.position) > 5 or
           (blackbox.position.x != nan and
            distance2d(blackbox.position, ship.position) > 6)) wait(1);
    while (distance2d(guy.position, ship.position) > 7) {
      message("Please get on the ship so we can blast off"); wait(10);
    }

    // Done
    ship.takeoff();
  }
```

All that's left for the `WheeledGrabber` to do is return to the ship, wait for the other bot to bring the Black Box, and then trigger liftoff. Note that if the astronaut isn't on board, the bot sends him a message and waits until he returns.

Note that in most missions, it is not actually necessary to have the bots on board, but the storyline would suggest they should be.

Next, we will see what the `TrackedGrabber` has to do:

```c++
  // Step 3: Get the black box and bring it back
  // (To be run by Tracked Grabber)
  public void step3() {
    // Wait until the other bot gives me power
    while(bot.energyCell == null) wait(1); wait(5);

    camerafocus(bot);
    message("Retrieving the black box");
    fetch(BlackBox);
    message("Bringing it to the ship");
    rgoto(ship); drop();
  }
```

This is fairly simple. When it first comes to life, the bot must wait for a power cell to be dropped on it. Then it waits an extra 5 seconds to give the `WheeledGrabber` time to get out of the way.

Claiming the camera focus (for the interest of those watching), it then goes to retrieve the Black Box. Returning to the ship, it drops the box. Done!

That just leaves the convenience methods to discuss:

```c++
  //-------------------------------------------------------------
  // Convenience methods
  //-------------------------------------------------------------

  // Indicates if the given research type is done yet
  public boolean researchDone(int researchType) {
    int researchDone = getresearchdone();
    return ((researchDone & researchType) != 0);
  }

  // Finds and grabs the nearest item of the given category
  public void fetch(int category) {
    // Already carrying it (or something else)?
    if (bot.load != null) {
      if (bot.load.category == category) return;
      else drop();
    }

    // Find it and grab it
    object item = radar(category);
    rgoto(item); grab();
  }
}
```

I think these are fairly straightforward. `researchDone()` hides the rather un-obvious way that the built-in `getresearchdone()` method returns its results (as a bit field-- see the documentation).

`fetch()` is a tad bit more interesting. It first checks if the bot is already carrying the desired resource type or something else, and reacts accordingly. If not, it finds the item, goes to it, and grabs it.

## Quirks in built-in `goto`

If you're sharp-eyed, you probably noticed I'm using a new function `rgoto()` instead of the built-in `goto()` function. Why is that?

Well, in previous versions of Colobot, and still now to a lesser degree, `goto` can be very quirky and unreliable. There are many edge cases where it fails to go to the desired location, gets stuck along the way, or wanders around aimlessly. In previous versions of Colobot, the `TrackedGrabber` could not directly `goto()` the Black Box. I had to write code like the following:

```c++
    // NOTE: this is old code, not a part of today's solution!
    //
    // These instructions were needed because for some reason the bot
    // didn't know how to go directly to the black box and grab it
    //
    // Go to a point 3m in front of the black box...
    point p;
    p.x = blackbox.position.x + cos(blackbox.orientation) * 3;
    p.y = blackbox.position.y + sin(blackbox.orientation) * 3;
    goto(p);
    // ...then turn toward it, and approach it with motor()...
    turn(direction(blackbox.position));
    motor(1,1); wait(2); motor(0,0); grab();
```

That worked, but through further experience with other levels, I decided to write my own wrapper for `goto()` to handle some of these quirks. At first I called it `mygoto()` but later settled on `rgoto()` (for "retry goto") because it's short. The code for it is below:

```c++
// My goto wrapper, which tries several times ("retry goto")
//
// This is helpful, because sometimes goto is buggy and refuses to go
// to a perfectly accessible location or item.
//
// NOTE: Pass the "item" you want to approach, not "item.position"
public int rgoto(object item) {
  errmode(0);
  int tries = 0, retval;

  // Sometimes, trying from a different start point works
  while((retval = goto(item.position)) != 0) {
    if (tries++ > 5) break;
    turn(rand() * 180 - 90);
    move((rand() - 0.3) * tries * 2);
  }

  // If not, try going to a point facing the item, from four different
  // directions if necessary. (This is why we want the item, not its
  // position, so we can use its orientation.)
  if (retval != 0) {
    point p;
    for (float angle = 0; angle < 360; angle += 90) {
      p.x = item.position.x + cos(item.orientation + angle) * 3;
      p.y = item.position.y + sin(item.orientation + angle) * 3;
      if (angle == 270) errmode(1); // last try
      if ((retval = goto(p)) == 0) break;
    }

    // ok, try approaching it with the motor
    turn(direction(item.position));
    motor(1,1); wait(2); motor(0,0);
  }

  errmode(1);
  return retval;
}
```

First, it sets `errmode(0)` which tells Colobot not to stop the program if it encounters an error. Then, when it calls built-in `goto()` it receives back its return value. If `goto()` succeeds, it returns `0`; if it fails, it returns something non-zero. In that case, my code tries starting out from random locations nearby.

If it exhausts a certain number of tries and still does not get there, it tries changing the destination. It changes the destination to a point 3 meters away from the desired target object, trying once from each of 4 different directions. If that succeeds, it turns toward the target item and uses `motor()` to move directly toward it.

This handled a large number of quirks in the original `goto()`. I noticed that in the current version of Colobot, they don't seem to occur for this level, but I kept the routine anyways. It gets drastically expanded in future versions for future levels.

## Conclusion

This is just my approach. With all of my walkthroughs, I am proud to present what I feel is one elegant solution. However, there are so many ways to solve a problem, and I am always curious to see how others do it. I love to learn from others, as I have already done over and over, whether with Colobot or with coding in general.

Do you have any interesting ways you solved this level, or any of the problems it presented? Please share in the comments below!

**Comment by SmileSick** [comment]
Very informative.  There are several questions to this.  Why is it so meticulous to prescribe these methods if they will not work in this form in other missions?  They will still need to be edited.  I like the approach to versatility, and I am guided by the same.  True, during the passage,  I have to edit my functions constantly. With a cursory review, I noticed that TrackedGrabber is waiting 5 seconds after receiving the battery.  Why not connect a radar for this. Example, while(radar(Any, 180,10,0,7)!=null) wait(1); Sorry for my English language

**Reply by cbot-coder** [reply]
As you rightly point out, these methods have to be adapted in almost the same form for the next mission, so at some point it makes sense to refactor them into a more reusable, parameterized form. I'm about to post my next walkthrough for Moon 1, where I discuss how the approach begins to show its limitations, and the beginnings of some ideas for improvements (which I didn't develop until some of the Tropica levels, and I keep refining).

You make a good point about the delay on the `TrackedGrabber`. It could try looking behind itself... I hadn't thought of that. I would love to see your completed implementation for that! You might need to do something to avoid picking up the `BotFactory`, but I think changing the radar's minimum range to about 2-3 meters should help.

Thanks for taking the time to comment!

**Reply by SmileSick** [reply]
My programs are in my repository. But i have not commented my code. I wrote the code while playing the game on my YouTube channel. I tried to keep the code as simple as possible(I'm not programmer).  The functions were rewritten during the passage. I followed the rule: one mission, one program.  Even if there are several bots at the level.  I used the resources to a minimum and went through the levels as if they were exercises with certain data.  I tried to exclude precise indications of coordinates and waiting time. Until I reached the planet volcano.

**Reply by cbot-coder** [reply]
Your approach sounds like it makes a lot of sense at a high level. I also try not to have hard-coded coordinates, etc. Waiting time is not a big deal for me, as I'm not (initially) trying to speed-walk the bots through. My goal is to get them to solve it reliably on their own. But speed may become the next challenge once I get through the levels. (So far I'm only up to Saari.)

I'll take a look at your repository. Thanks!

**Reply by SmileSick** [reply]
For speed I did speedrun with produce/delete functions. Just for fun.

**Reply by cbot-coder** [reply]
What's your YouTube channel?

**Reply by SmileSick** [reply]
https://youtube.com/user/Vladimirsville
Warning: russian language)))
